renderlib/RotateTool.cpp

```diff
@@ -0,0 +1,367 @@
+        angle = getDraggedAngle(camFrame.vz, p, l, l0, l1);
+        motion = glm::angleAxis(angle, vN);
+
+      } break;
+      case RotateTool::kRotateView: // constrained to rotate about view direction
+      {
+        // find angle between (p,l0) and (p,l1)
+        // using as N, the view direction
+        glm::vec3 vN = normalize(p - l);
+        glm::vec3 v0 = normalize(p - l0);
+        glm::vec3 v1 = normalize(p - l1);
+        angle = getSignedAngle(v0, v1, vN);
+        motion = glm::angleAxis(angle, vN);
+      } break;
+      case RotateTool::kRotate: // general tumble rotation
+        // use camera trackball algorithm
+        // scale pixels to radians of rotation (TODO)
+        float xRadians = -button.drag.x * dragScale * glm::two_pi<float>();
+        float yRadians = -button.drag.y * dragScale * glm::two_pi<float>();
+        angle = sqrtf(yRadians * yRadians + xRadians * xRadians);
+        glm::vec3 objectUpDirection = camFrame.vy * yRadians;
+        glm::vec3 objectSidewaysDirection = camFrame.vx * xRadians;
+
+        glm::vec3 moveDirection = objectUpDirection + objectSidewaysDirection;
+        glm::vec3 eye = l - p;
+        glm::vec3 axis = glm::normalize(glm::cross(moveDirection, eye));
+
+        motion = glm::angleAxis(angle, axis);
+
+        break;
+    }
+
+    // The variable motion is a world space vector of how much we moved the
+    // manipulator. Here we execute the action of applying motion to
+    // whatever is that we are moving...
+    // [...]
+    // Note that the motion is the total motion from drag origin.
+    // but we will update this per-frame!
+    origins.rotate(scene, motion);
+    m_rotation = motion;
+  }
+  if (button.action == Gesture::Input::Action::kRelease) {
+    if (!origins.empty()) {
+      // Make the edit final, for example by creating an undo action...
+      // [...]
+    }
+
+    // Consume the gesture.
+    gesture.input.reset(Gesture::Input::kButtonLeft);
+    origins.clear();
+    m_rotation = glm::angleAxis(0.0f, glm::vec3(0, 0, 1));
+  }
+}
+
+void
+RotateTool::draw(SceneView& scene, Gesture& gesture)
+{
+  const glm::vec2 resolution = scene.viewport.region.size();
+  // Move tool is only active if something is selected
+  if (!scene.anythingActive()) {
+    return;
+  }
+
+  // Re-read targetPosition because object may have moved by the manipulator action,
+  // or animation. Target 3x3 linear space is a orthonormal frame. Target
+  // position is where the manipulator should be drawn in space
+  if (origins.empty()) {
+    origins.update(scene);
+  }
+  AffineSpace3f target = origins.currentReference(scene, Origins::kNormalize);
+
+  glm::vec3 viewDir = (scene.camera.m_From - target.p);
+  LinearSpace3f camFrame = scene.camera.getFrame();
+
+  // Draw the manipulator to be at some constant size on screen
+  float scale = length(viewDir) * scene.camera.getHalfHorizontalAperture() * (s_manipulatorSize / resolution.x);
```

Comment:
Could you explain how this keeps the scale constant? I think I'm getting tripped up by the inclusion of the `getHalfHorizontalAperture()`, which is in radians.

Reply:
actually getHalfHorizontalAperture is not an angle.  It's half the horizontal width of the plane one unit away from the camera position, given a FoV angle.  (aka `tan(fov angle)` )